### DISCUSSION TITLE
Implementation challenges: Errors

### DISCUSSION THREAD

**Discussion by StefanoFalsetto-CKHIOD** [camaraproject/CallForwardingSignal#109] [General]
While the definition of the errors is detailed, the error cases are not so clear.

HTTP Error Code | Message | Comments
-- | -- | --
409 ABORTED | Concurrency conflict. | Which is the error case?
409 ALREADY_EXISTS | The resource that a client tried   to create already exists. | Which is the error case?
409 CONFLICT | A specified resource duplicate   entry found. | Which is the error case?
415 UNSUPPORTED_MEDIA_TYPE | The server refuses to accept the   request because the payload format is in an unsupported format. | Which is the error case?
422 UNIDENTIFIABLE_DEVICE | Phone number not defined. | should this error be raised when there is no phone number in request?   The 400 INVALID_ARGUMENT is not correct in that case?

Please help me answering to the questions into the "Comments" column.

**Comment by FabrizioMoggio** [comment]
These are defined by the CAMARA Guidelines: https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md

and in: https://github.com/camaraproject/Commonalities/blob/main/artifacts/CAMARA_common.yaml

Let's try to make some more reasoning in the context of the CFS API

**Comment by FabrizioMoggio** [comment]
According to the Guidelines the Following Error codes must be present:

for POST: 400, 401, 403, 409, 415, 429, 500, 503 

So we need 409 and 415

422 instead is used in 3Legs when the variable PhoneNumber differs from the related value in the AccessToken

**Reply by StefanoFalsetto-CKHIOD** [reply]
Thank you Fabrizio for clarifying to me the 422 Error.

I think the 409 ALREADY_EXISTS and 409 CONFLICT errors are out of scope since there are no resources to create (hence no duplication or conflict).
The 409 ABORTED could be used in case of an MNO is allowing only one request per API consumer per scope per MSISDN. But I think that's a limitation not useful to introduce.

What do you think?

415 error case is still unknown to me. In commonalties it is stated "should not happen"...

**Reply by FabrizioMoggio** [reply]
About 409, I just commented below. In my understanding we are sharing the same view

**Reply by FabrizioMoggio** [reply]
About 415, in my understanding it is used when the body is not in JSON format. Am I wrong (I'm not an expert)?

**Comment by FabrizioMoggio** [comment]
HTTP Error Code | Message | Comments
-- | -- | --
409 ABORTED | Concurrency conflict. | When the API is invoked by the same API Consumer for the same PhoneNumber and the previous API call is not yet completed. To avoid querying the network if the API Consumer loops.
409 ALREADY_EXISTS | The resource that a client tried   to create already exists. | For the CFS API this is as above
409 CONFLICT | A specified resource duplicate   entry found. | For the CFS API I don't see this happening 
415 UNSUPPORTED_MEDIA_TYPE | The server refuses to accept the   request because the payload format is in an unsupported format. | Which is the error case?
422 UNIDENTIFIABLE_DEVICE | Phone number not defined. | should this error be raised when there is no phone number in request?   The 400 INVALID_ARGUMENT is not correct in that case?

**Reply by FabrizioMoggio** [reply]
Anyway for 409 these are examples provided by the guidelines. Maybe we can delete some of those examples if not coherent with the CFS API (ALREADY_EXISTS, CONFLICT).

**Reply by StefanoFalsetto-CKHIOD** [reply]
I agree with you, I think at the moment we can remove the 409 from the specs.

**Reply by FabrizioMoggio** [reply]
one clarification. We can not remove 409, it is mandatory according to the Guideline, but we can just keep 409 ABORTED, removing the other ones. Is this possible?

**Reply by StefanoFalsetto-CKHIOD** [reply]
We can keep it in specifications but I think in our implementation we will never raise this error.

**Comment by FabrizioMoggio** [comment]
New Issue: https://github.com/camaraproject/CallForwardingSignal/issues/112

**Comment by bigludo7** [comment]
Hello
409 we can do what we want :)
For me fine to remove 409.

For the other rror, here the sum-up:
Mandatory Errors to be documented in CAMARA API Spec YAML:

Error status 401
Error status 403
Error status 429 TOO_MANY_REQUESTS (For rate limit control)

Errors not needed to be documented:

405 METHOD_NOT_ALLOWED
406 NOT_ACCEPTABLE
415 UNSUPPORTED_MEDIA_TYPE
 5xx errors

**Comment by FabrizioMoggio** [comment]
Let's remove 409.

**Comment by FabrizioMoggio** [comment]
@StefanoFalsetto-CKHIOD considering that many errors are no more mandatory we could remove some of them if not useful to our API. I would like to be coherent with your implementation and experience so, can you please identify the list of error you suggest to remove?

**Reply by StefanoFalsetto-CKHIOD** [reply]
@FabrizioMoggio this is a table with errors we defined as out of scope, and some notes:

Scenario | Error | Note
-- | -- | --
Inconsistent device | {     "status": 422,     "code":   "DEVICE_IDENTIFIERS_MISMATCH",     "message": "Provided device identifiers are not consistent."   } | There is it only one   identifier.
Concurrency | {     "status": 409,     "code":   "ABORTED",     "message":   "Concurrency conflict."   } | CKH is not using request queue. All the API   Calls can be performed in parallel.
Conflict | {     "status": 409,     "code": "CONFLICT",     "message": "A specified   resource duplicate entry found"   } | This   kind of service does not have specific state to be in conflict.
Already exists | {     "status": 409,     "code": "ALREADY_EXISTS",     "message": "The resource   that a client tried to create already exists"   } | This kind   of service does not create resources on server.
Unsupported   media type | {     "status": 415,     "code":   "UNSUPPORTED_MEDIA_TYPE",     "message": "The   server refuses to accept the request because the payload format is in an   unsupported format."   } | CKH   is not able to reproduce the problem. The description in commonalties is:    Payload   format of the request is in an unsupported format by the Server. Should not   happen


Actually, CKH not implemented other errors, depending on specific implementation (e.g., 501, not implemented)

**Comment by FabrizioMoggio** [comment]
here is a list of actions to be taken: https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/51413132/Analysis+of+Commonalities+0.5.0-alpha.1+changes#Analysis

**Comment by FabrizioMoggio** [comment]
- On 422 there are new guidelines (already implemented in the current version for Spring25). ok to remove DEVICE_IDENTIFIERS_MISMATCH
- As already agreed, ok to remove 409
- ok to remove 415

**Comment by FabrizioMoggio** [comment]
Implemented
